### PR TITLE
test/ClientActions : improve client actions tests 

### LINKED ISSUE #72: Improving Client Tests
The following scenarios will be added.

Scenario
- copy the log.toml file and safe-core.config file with unlimited mutations to a folder.
- use getAppStem and fetch the expected name for the app’s executable without extension. 
- set the additional search paths and initialize logging.
- create a session and set up the app container.
- perform some mutations.
- get the mDataInfo and perform some operations.
- get the account info and verify if puts have exceeded 1000.
- Verify if the output log file exists.
- getAppContainerName and verify.
- resetObjectCache and check the contents of permission handle and entries handle.
- refreshAccessInfo and fetch the previously accessible containers and verify that are not accessible.

Scenario
- Create a session
- Create, encode and send a container request to the authenticator with additional permissions.
- Get the containerMdInfo for the additional container and perform some operations.

### PR DESCRIPTION
**Pull request by vigneshwara044**
- Rename ClientTest to ClientActions
- Add additional tests to fetch custom config and initialise logging, access default containers, obtain additional container permissions, etc.
- Add new test helpers for handling IPC Request
closes #72

### CODE REVIEW

File: safe-app-android/src/main/java/net/maidsafe/api/Client.java [hunk 1]
```diff
@@ -35,7 +35,7 @@ public static void load(Context context) {
             Os.setenv("SAFE_MOCK_VAULT_PATH", path, true);
             Session.setAdditionalSearchPath(path).get();
         } catch (Exception e) {
-            throw new RuntimeException("Cannot set MockVault path");
+            throw new RuntimeException(e);
```
Comment (lionel1704) [inline safe-app-android/src/main/java/net/maidsafe/api/Client.java]: Revert this change please.

File: safe-app/src/test/java/net/maidsafe/api/ClientActions.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        String containersResp = TestHelper.handleIpcRequest(authenticator, request.getUri());
+        decodeResult = session.decodeIpcMessage(containersResp).get();
+        ContainerResponse containerResponse = (ContainerResponse) decodeResult;
+        Assert.assertEquals(containerResponse.getReqId(), request.getReqId());
+        session.refreshAccessInfo();
+
+        NativeHandle actionHandle = session.mDataEntryAction.newEntryAction().get();
+        String keyString = Helper.randomAlphaNumeric(LENGTH);
+        String valueString = Helper.randomAlphaNumeric(LENGTH);
+        session.mDataEntryAction.insert(actionHandle, keyString.getBytes(), valueString.getBytes()).get();
```
Comment (lionel1704) [inline safe-app/src/test/java/net/maidsafe/api/ClientActions.java]: You need to perform a mutation to check whether the additional permissions have been granted.

File: safe-app/src/test/java/net/maidsafe/api/MutableDataOperations.java [hunk 3]
```diff
@@ -353,7 +355,11 @@ public void shouldShareMutableData() throws Exception {
                     Helper.randomAlphaNumeric(LENGTH), Helper.randomAlphaNumeric(LENGTH)),
                     true, permissions, 1, 0);
         try {
-            Session appA = TestHelper.handleAuthReq(authenticator, authReq);
+            Request ipcMessage = Session.encodeAuthReq(authReq).get();
+            String encodedResponseReceived = TestHelper.handleIpcRequest(authenticator, ipcMessage.getUri());
```
Comment (lionel1704) [inline safe-app/src/test/java/net/maidsafe/api/MutableDataOperations.java]: `encodedAuthResponse`*

File: safe-app/src/test/java/net/maidsafe/api/MutableDataOperations.java [hunk 4]
```diff
@@ -381,7 +387,12 @@ public void shouldShareMutableData() throws Exception {
         }
 
         authReq.getApp().setId("net.maidsafe.app.two");
-        Session appB = TestHelper.handleAuthReq(authenticator, authReq);
+        Request ipcMessage = Session.encodeAuthReq(authReq).get();
+        String encodedResponseReceived = TestHelper.handleIpcRequest(authenticator, ipcMessage.getUri());
```
Comment (lionel1704) [inline safe-app/src/test/java/net/maidsafe/api/MutableDataOperations.java]: `encodedAuthResponse`*

File: safe-app/src/test/java/net/maidsafe/api/ClientActions.java [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+    // unlimited mutations set to true.
+    // Set the path to the temporary directory as an additional search path.
+    // Initialise logging to a Client.log file.
+    // Create a session and perform some operations.
+    // Verify that the log file exists and the content is not "null".
+    // Verify the increase in mutations done.
+    // Verify that the number of mutations available is always 1000.
+    @Test
+    public void shouldFetchCustomConfigAndInitialiseLogging() throws Exception {
+        File generatedDir = Helper.copyFiles();
```
Comment (lionel1704) [inline safe-app/src/test/java/net/maidsafe/api/ClientActions.java]: Can we rename this function to `copyConfigFiles`?

Comment (vigneshwara044) [reply safe-app/src/test/java/net/maidsafe/api/ClientActions.java]: ok

File: safe-app-android/src/androidTest/java/net/maidsafe/test/utils/Helper.java [hunk 6]
```diff
@@ -17,6 +26,43 @@
...
+    }
+
+    private static void copyFile(String fileName, File destinationDir, String appName) throws Exception {
+        InputStream inputStream = InstrumentationRegistry
+                                        .getContext()
+                                        .getAssets()
+                                        .open(fileName);
+        String destFileName = fileName;
+        if (appName != null) {
+            destFileName = appName + "." + destFileName;
```
Comment (krishnaIndia) [inline safe-app-android/src/androidTest/java/net/maidsafe/test/utils/Helper.java]: Isn't `StringBuilder` of better use here?

File: safe-app/src/test/java/net/maidsafe/api/ClientActions.java [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+        client.testSimulateDisconnect().get();
+    }
+
+    // Create a session with access to _public container.
+    // Fetch the list of container permissions and verify them.
+    // Fetch MDataInfo for the _public container and verify that mutations are allowed.
+    @Test
+    public void shouldHaveAccessToDefaultContainer() throws Exception {
+        ContainerPermissions[] permissions = new ContainerPermissions[1];
+        permissions[0] = new ContainerPermissions("_public", new PermissionSet(true,
```
Comment (krishnaIndia) [inline safe-app/src/test/java/net/maidsafe/api/ClientActions.java]: Can the container names be defined as constants? Also we are using empty string in a lot of places. Make it a constant too?

File: safe-app/src/test/java/net/maidsafe/api/ClientActions.java [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+        AppExchangeInfo appExchangeInfo = new AppExchangeInfo(APP_ID, "", Helper.randomAlphaNumeric(LENGTH),
+                                                Helper.randomAlphaNumeric(LENGTH));
+        AuthReq authReq = new AuthReq(appExchangeInfo, true, permissions, 1, 0);
+        Request ipcMessage = Session.encodeAuthReq(authReq).get();
+        String encodedResponse = TestHelper.handleIpcRequest(authenticator, ipcMessage.getUri());
+        DecodeResult decodeResult = Session.decodeIpcMessage(encodedResponse).get();
+        AuthResponse authResponse = (AuthResponse) decodeResult;
+        Session session = Session.connect(authReq.getApp().getId(), authResponse.getAuthGranted()).get();
+
+        ContainerPermissions[] newPermissions = new ContainerPermissions[1];
```
Comment (krishnaIndia) [inline safe-app/src/test/java/net/maidsafe/api/ClientActions.java]: Is it better to have it as list, keep appending to the list and convert to an array when needed? So that magic numbers can be avoided?
